### DISCUSSION TITLE
Authenticating with private NuGet feed?

### DISCUSSION THREAD

**Discussion by vsosinga** [renovatebot/renovate#10998] [Request Help] [answered]
I'm using a self-hosted Renovate setup in Azure DevOps and having some trouble to connecting to my private NuGet feed (also in DevOps).
I keep getting the following error: `Response status code does not indicate success: 401 (Unauthorized)`

Feed url: `https://pkgs.dev.azure.com/<org>/_packaging/<org>/nuget/v3/index.json`

Relevant lines from my config:
```
"hostRules": [
  {
    "hostType": "nuget",
    "username": "",
    "password": "***********",
    "matchHost": "pkgs.dev.azure.com"
  }
]
```
The password is a Personal Access Token generated in DevOps.

Debug log:
```
DEBUG: Executing command (repository=<repo>, branch=renovate/all-minor-patch)
        "command": "dotnet restore Project.csproj --force-evaluate --configfile /tmp/renovate/cache/others/nuget/8f1a7dfefd0e7797/nuget.config"
DEBUG: rawExec err (repository=<repo>, branch=renovate/all-minor-patch)
       "err": {
         "killed": false,
         "code": 1,
          "signal": null,
         "cmd": "dotnet restore Project.csproj --force-evaluate --configfile /tmp/renovate/cache/others/nuget/8f1a7dfefd0e7797/nuget.config",
         "stdout": "  Determining projects to restore...\n/usr/share/dotnet/sdk/5.0.302/NuGet.targets(131,5): error : Unable to load the service index for source https://pkgs.dev.azure.com/<org>/_packaging/<org>/nuget/v3/index.json. [/tmp/renovate/repos/azure/Project.csproj]\n/usr/share/dotnet/sdk/5.0.302/NuGet.targets(131,5): error :   Response status code does not indicate success: 401 (Unauthorized). [/tmp/renovate/repos/azure/Project.csproj]\n",
         "stderr": "",
         "message": "Command failed: dotnet restore Project.csproj --force-evaluate --configfile /tmp/renovate/cache/others/nuget/8f1a7dfefd0e7797/nuget.config\n",
         "stack": "Error: Command failed: dotnet restore Project.csproj --force-evaluate --configfile /tmp/renovate/cache/others/nuget/8f1a7dfefd0e7797/nuget.config\n\n    at ChildProcess.exithandler (child_process.js:319:12)\n    at ChildProcess.emit (events.js:375:28)\n    at ChildProcess.emit (domain.js:470:12)\n    at maybeClose (internal/child_process.js:1055:16)\n    at Process.ChildProcess._handle.onexit (internal/child_process.js:288:5)"
       }
```
Connecting directly to the feed using curl with basic authentication works without problems.
I also tried the setup in this discussion: https://github.com/renovatebot/renovate/discussions/8576, but without success.

Any suggestions?

**Comment by rarkins** [comment, marked as answer]
Seems to follow our docs here: https://docs.renovatebot.com/private-modules/#nuget

However maybe the empty string username is causing it not to be executed. Can you confirm exactly how you verified the curl command? e.g. did you base64 encrypt the string like `:mypassword`?

**Reply by vsosinga** [reply]
I used Postman to manually test the endpoint with an empty username and the password. The authorization header it generates, also for the curl command, is equal to `:mypassword` base64 encoded indeed.

In the docs you reference it is stated "For those found, a command similar to the following is run: `dotnet nuget add source ${registryInfo.feedUrl} --configfile ${nugetConfigFile} --username ${username} --password ${password} --store-password-in-clear-text`"
I can't find that line in my debug log however, only this one without username/password:
`dotnet nuget add source https://pkgs.dev.azure.com/<org>/_packaging/<org>/nuget/v3/index.json --configfile /tmp/renovate/cache/others/nuget/8f1a7dfefd0e7797/nuget.config --name <org>`

**Reply by rarkins** [reply]
I can confirm that we would suppress this command if username is an empty string. Are you able to confirm if We should run `dotnet nuget add source ... --username "" --password "mpypassword"` or if it should drop the username completely like `dotnet nuget add source ...  --password "mpypassword"`? After that I can apply a fix

**Reply by vsosinga** [reply]
just confirmed: `username` is a required parameter and has to have a value,  `""` doesn't work. The value doesn't matter for the authentication with DevOps, so the following would be correct:
`dotnet nuget add source ... --username "none" --password "mpypassword"`

I also tried to set a dummy `username` in my `hostRules` (`"username": "none"`), but I still don't see the username & password being set in my debug log.

**Reply by rarkins** [reply]
The logic is here: https://github.com/renovatebot/renovate/blob/983c0626e182796e081639cd8cab0a777138e3af/lib/manager/nuget/artifacts.ts#L37-L55

It relies on there being a list of registries returned by this command: https://github.com/renovatebot/renovate/blob/983c0626e182796e081639cd8cab0a777138e3af/lib/manager/nuget/util.ts#L33

If you'd like to submit a PR with more debug logging (e.g. listing the configured registries and whether or not authentication was found) then we'd accept it.

**Reply by vsosinga** [reply]
Thanks for the reference to the logic. If I understand it correctly the `hostRules` entry is matched based on the full url of the feed?
I changed my settings to the following:
```
"hostRules": [{
	"hostType": "nuget",
	"url": "https://pkgs.dev.azure.com/<org>/_packaging/<org>/nuget/v3/index.json",
	"username": "none",
	"password": "***********"
}]
```
And the feed is now being authenticated:
`dotnet nuget add source ... --username none --password **redacted** --store-password-in-clear-text`
🥳

**Reply by rarkins** [reply]
Great! Could you submit a PR expanding the documentation in https://github.com/renovatebot/renovate/blob/main/docs/usage/private-modules.md for future users? e.g. the full host match etc

**Reply by pim-simons** [reply]
Thanks for this info as this got me very close to getting this working. 
However setting the hostRules url to the entire entry (`https://pkgs.dev.azure.com/<org>/_packaging/<org>/nuget/v3/index.json`) did not work for me as the tracing still showed unauthorized while getting this url: `https://pkgs.dev.azure.com/<org>/_packaging/<org>/nuget/v3/registrations2-semver2/<package>/index.json`

Changing the settings to this worked for me:
```
"hostRules": [{
	"hostType": "nuget",
	"url": "https://pkgs.dev.azure.com/<org>",
	"username": "none",
	"password": "***********"
}]
```

Hopefully this helps someone.

**Reply by vsosinga** [reply]
Hmmz, I did some more testing and digging in the renovate sources and I think my understanding of the logic was incorrect.
The `hostRules` entry isn't matched on the `url`, but on `matchHost` as is stated in the docs. What probably happens when you use `url` is that it applies the `username` and `password` to all NuGet feeds. That also made that the default feed stopped working for me.
I now changed my settings to this:
```
  hostRules: [
    {
      hostType: "nuget",
      matchHost: "https://pkgs.dev.azure.com/<org>/",
      username: "none",
      password: <PAT>
    }
  ]
```
And now both my private feed and the default feed are working.

Only problem I currently have is that 3 `.csproj` files are being updated, but only one of them has gets an updated `packages.lock.json`
